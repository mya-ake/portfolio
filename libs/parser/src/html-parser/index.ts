export * from './html-parser';
export * from './utils';
