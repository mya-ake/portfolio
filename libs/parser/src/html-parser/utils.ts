import type { Node, TagNode, TextNode } from './html-parser';

export const isTagNode = (node: Node): node is TagNode => {
  return node.nodeType === 'tag';
};

export const isTextNode = (node: Node): node is TextNode => {
  return node.nodeType === 'text';
};

export const findTargetTagNode = (
  targetTagName: string,
  nodes: Node[],
): TagNode | undefined => {
  return nodes.filter(isTagNode).find((node) => {
    return node.tagName === targetTagName;
  });
};

export const hasTargetTagNode = (
  targetTagName: string,
  nodes: Node[],
): boolean => {
  return !!findTargetTagNode(targetTagName, nodes);
};
