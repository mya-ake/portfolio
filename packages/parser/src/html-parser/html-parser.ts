/* eslint-disable @typescript-eslint/no-explicit-any */
import parse5 from 'parse5';

type NodeType = 'tag' | 'text';
type Attributes = Record<string, string>;

export type TagNode = {
  nodeType: 'tag';
  tagName: string;
  attrs: Attributes;
  childNodes: Node[];
};

export type TextNode = {
  nodeType: 'text';
  content: string;
};

export type Node = TagNode | TextNode;

const getNodeType = (nodeName: string): NodeType => {
  switch (nodeName) {
    case '#text':
      return 'text';
    default:
      return 'tag';
  }
};

const createTagNode = (childNode: parse5.Element): TagNode => ({
  nodeType: 'tag',
  tagName: childNode.tagName,
  attrs: childNode.attrs.reduce<Attributes>((attrs, attrItem) => {
    attrs[attrItem.name] = attrItem.value;
    return attrs;
  }, {}),
  childNodes: convertNodes((childNode as any).childNodes ?? []),
});

const createTextNode = (childNode: parse5.TextNode): TextNode => ({
  nodeType: 'text',
  content: childNode.value,
});

const convertNode = (childNode: parse5.ChildNode): Node => {
  const nodeType = getNodeType(childNode.nodeName);
  switch (nodeType) {
    case 'tag':
      return createTagNode(childNode as parse5.Element);
    case 'text':
      return createTextNode(childNode as parse5.TextNode);
  }
};

const convertNodes = (childNodes: parse5.ChildNode[]): Node[] =>
  childNodes.map(convertNode);

export const parseHtml = (html: string): Node[] => {
  const documentFragment = parse5.parseFragment(html);
  return convertNodes(documentFragment.childNodes);
};
