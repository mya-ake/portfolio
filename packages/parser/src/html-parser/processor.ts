export const removeBackslashOfHTMLAttribute = (html: string): string =>
  html.replace(/\\"/g, '"');
