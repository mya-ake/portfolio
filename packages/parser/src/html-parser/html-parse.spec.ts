import { parseHtml } from './html-parser';

describe('parseHtml', () => {
  describe('single child', () => {
    it('no attrs', () => {
      const result = parseHtml(`<p>text</p>`);
      expect(result).toEqual([
        {
          nodeType: 'tag',
          tagName: 'p',
          attrs: [],
          childNodes: [
            {
              nodeType: 'text',
              content: 'text',
            },
          ],
        },
      ]);
    });

    it('has attrs', () => {
      const result = parseHtml(
        `<button class="some-class" type="button" disabled>text</button>`,
      );
      expect(result).toEqual([
        {
          nodeType: 'tag',
          tagName: 'button',
          attrs: [
            {
              name: 'class',
              value: 'some-class',
            },
            {
              name: 'type',
              value: 'button',
            },
            {
              name: 'disabled',
              value: '',
            },
          ],
          childNodes: [
            {
              nodeType: 'text',
              content: 'text',
            },
          ],
        },
      ]);
    });
  });
});
