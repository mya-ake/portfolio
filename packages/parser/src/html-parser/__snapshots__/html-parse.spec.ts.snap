// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseHtml deep children list 1`] = `
Array [
  Object {
    "attrs": Object {},
    "childNodes": Array [
      Object {
        "attrs": Object {},
        "childNodes": Array [
          Object {
            "content": "item 1",
            "nodeType": "text",
          },
        ],
        "nodeType": "tag",
        "tagName": "li",
      },
      Object {
        "attrs": Object {},
        "childNodes": Array [
          Object {
            "content": "item 1",
            "nodeType": "text",
          },
        ],
        "nodeType": "tag",
        "tagName": "li",
      },
    ],
    "nodeType": "tag",
    "tagName": "ul",
  },
]
`;

exports[`parseHtml deep children some content 1`] = `
Array [
  Object {
    "attrs": Object {},
    "childNodes": Array [
      Object {
        "content": "no wrapped",
        "nodeType": "text",
      },
      Object {
        "attrs": Object {},
        "childNodes": Array [
          Object {
            "content": "item 1",
            "nodeType": "text",
          },
        ],
        "nodeType": "tag",
        "tagName": "p",
      },
      Object {
        "attrs": Object {
          "alt": "",
          "src": "/static/sample.png",
        },
        "childNodes": Array [],
        "nodeType": "tag",
        "tagName": "img",
      },
    ],
    "nodeType": "tag",
    "tagName": "div",
  },
]
`;

exports[`parseHtml web components list 1`] = `
Array [
  Object {
    "attrs": Object {},
    "childNodes": Array [
      Object {
        "content": "text",
        "nodeType": "text",
      },
    ],
    "nodeType": "tag",
    "tagName": "custom-p",
  },
]
`;
